aiohue/v2/controllers/base.py

```diff
@@ -0,0 +1,234 @@
+from aiohue.v2.models.device import Device
+
+from ...util import dataclass_to_dict, update_dataclass, NoneType
+from ..models.clip import CLIPResource, parse_clip_resource
+from ..models.resource import ResourceTypes
+from .events import EventCallBackType, EventType
+
+if TYPE_CHECKING:
+    from .. import HueBridgeV2
+
+
+EventSubscriptionType = Tuple[
+    EventCallBackType,
+    "Tuple[str] | None",
+    "Tuple[EventType] | None",
+]
+
+
+class BaseResourcesController(Generic[CLIPResource]):
+    """Holds and manages all items for a specific Hue resource type."""
+
+    item_type = ResourceTypes.UNKNOWN
+
+    def __init__(self, bridge: "HueBridgeV2") -> None:
+        """Initialize instance."""
+        self._bridge = bridge
+        self._items: Dict[str, CLIPResource] = {}
+        self._logger = bridge.logger.getChild(self.item_type.value)
+        self._subscribers: List[EventSubscriptionType] = []
+
+    @property
+    def items(self) -> List[CLIPResource]:
+        """Return all items for this resource."""
+        return list(self._items.values())
+
+    async def initialize(self, initial_data: List[dict] | None) -> None:
+        """
+        Initialize controller by fetching all items for this resource type from bridge.
+
+        Must be called only once. Any updates will be retrieved by events.
+        Optionally provide the initial data.
+        """
+        if initial_data is None:
+            endpoint = f"clip/v2/resource/{self.item_type.value}"
+            initial_data = await self._bridge.request("get", endpoint)
+        item_count = 0
+        for item in initial_data:
+            if ResourceTypes(item["type"]) != self.item_type:
+                continue
+            item_count += 1
+            res: CLIPResource = parse_clip_resource(item)
+            self._items[res.id] = parse_clip_resource(item)
+        # subscribe to item updates
+        self._bridge.events.subscribe(
+            self._handle_event, resource_filter=self.item_type
+        )
+        self._logger.debug("fetched %s items", item_count)
+
+    def subscribe(
+        self,
+        callback: EventCallBackType,
+        id_filter: str | Tuple[str] | None = None,
+        event_filter: EventType | Tuple[EventType] | None = None,
+    ) -> Callable:
+        """
+        Subscribe to status changes for this resource type.
+
+        Parameters:
+            - `callback` - callback function to call when an event emits.
+            - `id_filter` - Optionally provide a resource ID to filter events for.
+            - `event_filter` - Optionally provide an EventType as filter.
+
+        Returns:
+            function to unsubscribe.
+        """
+        if not isinstance(event_filter, (NoneType, tuple)):
+            event_filter = (event_filter,)
+        if not isinstance(id_filter, (NoneType, tuple)):
+            id_filter = (id_filter,)
+        subscription = (callback, id_filter, event_filter)
```

Comment:
Most entities will subscribe to a single `id`. By storing it as tuples, it's no longer possible to do O(1) lookups but instead we're iterating the whole list of all handlers if an event comes in. 

I would suggest that we keep track of 2 subscriber dictionaries. One is `single_id_subscribers` and it's a dictionary keyed by `id`.

Reply:
Yeah, I was thinking in that same direction a few days ago but was afraid of over complicating stuff.
But you're right, iterating the entire list is a waste of resources. I'll adjust it

Reply:
I've adjusted it now